File: Ruby/DeepSpaceRuby/lib/SpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStationToUI'
require_relative 'Transformation'

module Deepspace
  class SpaceStation
    @@MAXFUEL = 100
    @@SHIELDLOSSPERUNITSHOT = 0.1

    attr_reader :ammoPower, :fuelUnits, :hangar, :name, :nMedals, :pendingDamage, :shieldBoosters, :shieldPower, :weapons

    def initialize (n, ap, fu, sp, nm=0, sb=[], h=nil, w=[], pd=nil)
      @name = n
      @ammoPower = ap
      assignFuelValue(fu)
      @nMedals = nm
      @shieldBoosters = sb
      @hangar = h
      @weapons = w
      @shieldPower = sp
      @pendingDamage = pd
    end

    def self.newStation(n,s)
      new(n,s.ammoPower,s.fuelUnits,s.shieldPower)
    end

    def self.newCopy(station)
      stationCopy = SpaceStation.new(station.name, station.ammoPower, station.fuelUnits, station.shieldPower, station.nMedals, station.shieldBoosters, station.hangar, station.weapons, station.pendingDamage)
      stationCopy
    end

    private
    def assignFuelValue (f)
      if f > @@MAXFUEL then
        @fuelUnits = @@MAXFUEL
      else
        @fuelUnits = f
      end
    end

    def cleanPendingDamage
      if @pendingDamage != nil && @pendingDamage.hasNoEffect then
        @pendingDamage = nil
      end
    end

    public
    def cleanUpMountedItems
      i = 0

      while i < @weapons.length do
        if @weapons[i].uses <= 0 then
          @weapons.delete_at(i)
        end

        i +=1
      end

      i = 0

      while i < @shieldBoosters.length do
        if @shieldBoosters[i].uses <= 0 then
          @shieldBoosters.delete_at(i)
        end

        i += 1
      end
    end

    def discardHangar
      @hangar = nil
    end

    def discardShieldBooster (i)
      if (i >= 0 && i < @shieldBoosters.size) then
        @shieldBoosters.delete_at(i)

        if @pendingDamage != nil then
          @pendingDamage.discardShieldBooster
          cleanPendingDamage
        end
      end
    end

    def discardShieldBoosterInHangar (i)
      if @hangar != nil then
        @hangar.removeShieldBooster(i)
      end
    end

    def discardWeapon (i)
      if (i >= 0 && i < @weapons.size) then
        w = @weapons.delete_at(i)

        if @pendingDamage != nil then
          @pendingDamage.discardWeapon(w)
          cleanPendingDamage
        end
      end
    end

    def discardWeaponInHangar (i)
      if @hangar != nil then
        @hangar.removeWeapon(i)
      end
    end

    def fire
      factor = 1

      if !@weapons.empty? then
        for i in 0..@weapons.length-1
          w = @weapons.at(i)
          factor *= w.useIt
        end
      end

      factor *= @ammoPower

      return factor
    end

    def speed
      return (@fuelUnits.to_f/@@MAXFUEL.to_f)
    end

    def getUIversion
      return SpaceStationToUI.new(self)
    end

    def mountShieldBooster (i)
      if @hangar != nil then
        shield = @hangar.removeShieldBooster(i)

        if shield != nil then
          @shieldBoosters.push(shield)
        end
      end
    end

    def mountWeapon (i)
      if @hangar != nil then
        weapon = @hangar.removeWeapon(i)

        if weapon != nil then
          @weapons.push(weapon)
        end
      end
    end

    def move
      @fuelUnits = @fuelUnits.to_f - (@fuelUnits.to_f * speed() )

      if @fuelUnits < 0 then
        @fuelUnits = 0
      end
    end

    def protection
      factor = 1

      if !@shieldBoosters.empty? then
        for i in 0..@shieldBoosters.length-1
          s = @shieldBoosters.at(i)
          factor *= s.useIt
        end
      end

      factor *= @shieldPower

      return factor
    end

    def receiveHangar (n)
      if @hangar == nil then
        @hangar = n
      end
    end

    def receiveShieldBooster (s)
      if @hangar != nil then
        return @hangar.addShieldBooster(s)
      else
        return false
      end
    end

    def receiveShot (shot)
      myProtection = protection

      if myProtection >= shot then
        @shieldPower -= @@SHIELDLOSSPERUNITSHOT * shot
        @shieldPower = [0.0, @shieldPower].max

        return ShotResult::RESIST
      else
        @shieldPower = 0.0

        return ShotResult::DONOTRESIST
      end
    end

    def receiveSupplies (s)
      @ammoPower += s.ammoPower()
      @fuelUnits += s.fuelUnits()
      @shieldPower += s.shieldPower()

      if @fuelUnits > @@MAXFUEL then
        @fuelUnits = @@MAXFUEL
      end
    end

    def receiveWeapon (w)
      if @hangar != nil then
        return @hangar.addWeapon(w)
      else
        return false
      end
    end

    def setLoot (loot)
      dealer = CardDealer.instance
      h = loot.nHangars

      if h > 0 then
        @hangar = dealer.nextHangar
        receiveHangar(@hangar)
      end

      elements = loot.nSupplies

      for i in 0..(elements - 1)
        sup = dealer.nextSuppliesPackage
        receiveSupplies(sup)
      end

      elements = loot.nWeapons

      for j in 0..(elements - 1)
        weap = dealer.nextWeapon
        receiveWeapon(weap)
      end

      elements = loot.nShields

      for k in 0..(elements - 1)
        sh = dealer.nextShieldBooster
        receiveShieldBooster(sh)
      end

      medals = loot.nMedals

      @nMedals += medals

      if loot.spaceCity
        return Transformation::SPACECITY
      elsif loot.efficient
        return Transformation::GETEFFICIENT
      else
        return Transformation::NOTRANSFORM
      end
    end

    def setPendingDamage (d)
      @pendingDamage = d.adjust(@weapons, @shieldBoosters)
    end

    def validState
      cleanPendingDamage

      if @pendingDamage == nil then
        return true
      else
        return false
      end
    end

    def to_s
      cad = "maxFuel: #{@@MAXFUEL}, shielLossPerUnitShot: #{@@SHIELDLOSSPERUNITSHOT}"
      cad += "\nammoPower: #{@ammoPower}, fuelUnits: #{@fuelUnits}, name: #{@name}, nMedals: #{@nMedals}"
      cad += "\nweapons: "

      @weapons.each{ |w|
        cad += w.to_s
      }

      cad += "\nshieldBoosters: "

      @shieldBoosters.each{ |sB|
        cad += sB.to_s
      }

      cad += "\nhangar: #{@hangar.to_s}"
      cad += "\nshieldPower: #{@shieldPower}, pendingDamage: #{@pendingDamage.to_s}"

      return cad
    end

  end
end

File: Java/DeepspaceJava/deepspace/src/deepspace/SpaceStation.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author paula
 */
public class SpaceStation implements SpaceFighter {
    private static final int MAXFUEL = 100;
    private static final float SHIELDLOSSPERUNITSHOT = 0.1f;
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private ArrayList<ShieldBooster> shieldBoosters = new ArrayList<>();
    private Hangar hangar;
    private ArrayList<Weapon> weapons = new ArrayList<>();
    private float shieldPower;
    private Damage pendingDamage;

    SpaceStation (String n, SuppliesPackage s){
        ammoPower = s.getAmmoPower();
        fuelUnits = s.getFuelUnits();
        name = n;
        nMedals = 0;
        hangar = null;
        shieldPower = s.getShieldPower();
        pendingDamage = null;
    }
    
    SpaceStation (SpaceStation station){
        this(station.name, new SuppliesPackage(station.ammoPower, station.fuelUnits, station.shieldPower));
        nMedals = station.nMedals;
        weapons = new ArrayList(station.weapons);
        shieldBoosters = new ArrayList(station.shieldBoosters);
        hangar = null;
        
        if (station.hangar != null)
            hangar = new Hangar(station.hangar);
        
        pendingDamage = null;
  
        if(station.pendingDamage != null)
            pendingDamage = station.pendingDamage.copy();
    }
    
    public float getAmmoPower (){
        return ammoPower;
    }
    
    public float getFuelUnits (){
        return fuelUnits;
    }
    
    public Hangar getHangar (){
        return hangar;
    }
    
    public String getName (){
        return name;
    }
    
    public int getNMedals (){
        return nMedals;
    }
    
    public Damage getPendingDamage (){
        return pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters (){
        return shieldBoosters;
    }
    
    public float getShieldPower (){
        return shieldPower;
    }
    
    public ArrayList<Weapon> getWeapons (){
        return weapons;
    }

    private void assignFuelValue (float f){
        fuelUnits = f;
    }

    private void cleanPendingDamage (){
        if (pendingDamage.hasNoEffect())
            pendingDamage = null;
    }

    public void cleanUpMountedItems (){
        int i;
        
        if (weapons != null){
            i = 0;
            while (i<weapons.size()){
                if (weapons.get(i).getUses() == 0){
                    weapons.remove(i);
                }
                else{
                    i++;
                }
            }
        }
        
        i = 0;
        while (i<shieldBoosters.size()){
            if (shieldBoosters.get(i).getUses() == 0){
                shieldBoosters.remove(i);
            }
            else{
                i++;
            }
        }
    }

    public void discardHangar (){
        hangar = null;
    }

    public void discardShieldBooster (int i){
        if (i >= 0 && i< shieldBoosters.size()){
            ShieldBooster s = shieldBoosters.remove(i);

            if (pendingDamage != null){
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }
    }

    public void discardShieldBoosterInHangar (int i){
        if (hangar != null)
            hangar.removeShieldBooster(i);
    }

    public void discardWeapon (int i){
        if (i >= 0 && i< weapons.size()){
            Weapon w = weapons.remove(i);

            if (pendingDamage != null){
                pendingDamage.discardWeapon(w);
                cleanPendingDamage();
            }
        }
    }
    

    public void discardWeaponInHangar (int i){
        if (hangar != null)
            hangar.removeWeapon(i);
    }

    public float fire (){
        float factor = 1;
        Weapon w;
        
        if (!weapons.isEmpty()){
            for (int i=0; i<weapons.size(); i++){
                w = weapons.get(i);
                factor *= w.useIt();
            }
        }
        
        factor *= ammoPower;
        
        return factor;
    }

    public float getSpeed (){
        return fuelUnits/MAXFUEL;
    }

    public SpaceStationToUI getUIversion (){
        return new SpaceStationToUI(this);
    }

    public void mountShieldBooster (int i){
        if (hangar != null && i >= 0) {
            ShieldBooster shield = hangar.removeShieldBooster(i);

            if (shield != null)
                shieldBoosters.add(shield);
        }
    }

    public void mountWeapon (int i){
        if (hangar != null && i >= 0){
            Weapon weapon = hangar.removeWeapon(i);

            if (weapon != null)
                weapons.add(weapon);
        }
    }

    public void move (){
        fuelUnits = fuelUnits - (fuelUnits * getSpeed());

        if (fuelUnits < 0)
            fuelUnits = 0;
    }

    public float protection (){
        float factor = 1;
        ShieldBooster s;
        
        if (!shieldBoosters.isEmpty()){
            for (int i=0; i<shieldBoosters.size(); i++){
                s = shieldBoosters.get(i);
                factor *= s.useIt();
            }
        }
        
        factor *= ammoPower;
        
        return factor;
    }

    public void receiveHangar (Hangar h){
        if (hangar == null)
            hangar = h; 
    }

    public boolean receiveShieldBooster (ShieldBooster s){
        if (hangar != null)
            return hangar.addShieldBooster(s);
        else
            return false;
    }

    public ShotResult receiveShot (float shot){
        float myProtection = protection();
        
        if (myProtection >= shot){
            shieldPower -= SHIELDLOSSPERUNITSHOT * shot;
            
            if (shieldPower < 0.0f)
                shieldPower = 0.0f;
            
            return ShotResult.RESIST;          
        }
        else{
            shieldPower = 0.0f;
            
            return ShotResult.DONOTRESIST;
        }
    }   

    public void receiveSupplies (SuppliesPackage s){
        ammoPower += s.getAmmoPower();
        fuelUnits += s.getFuelUnits();
        shieldPower += s.getShieldPower();
    }

    public boolean receiveWeapon (Weapon w){
        if (hangar != null)
            return hangar.addWeapon(w);
        else
            return false;
    }

    public Transformation setLoot (Loot loot){
        CardDealer dealer = CardDealer.getInstance();
        int h = loot.getNHangars();
        
        if (h > 0){
            hangar = dealer.nextHangar();
            receiveHangar(hangar);
        }
        
        int elements = loot.getNSupplies();
        
        for (int i=0; i<elements; i++){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }
        
        elements = loot.getNWeapons();
        
        for (int i=0; i<elements; i++){
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }
        
        elements = loot.getNShields();
        
        for (int i=0; i<elements; i++){
            ShieldBooster sh = dealer.nextShieldBooster();
            receiveShieldBooster(sh);
        }
        
        int medals = loot.getNMedals();
        
        nMedals += medals;
        
        if (loot.getSpaceCity())
            return Transformation.SPACECITY; //sintaxis??
        else if (loot.getEfficient())
            return Transformation.GETEFFICIENT;
        else
            return Transformation.NOTRANSFORM;
    }

    public void setPendingDamage (Damage d){
        pendingDamage = d.adjust(weapons, shieldBoosters);
    }

    public boolean validState (){
        if (pendingDamage == null || pendingDamage.hasNoEffect())
            return true;
        else
            return false;
    }
}
